geode-membership/src/main/java/org/apache/geode/distributed/internal/membership/gms/GMSMembership.java

```diff
@@ -273,6 +273,9 @@ boolean isDistributionMessage() {
   /**
    * Members that have sent a shutdown message. This is used to suppress suspect processing that
    * otherwise becomes pretty aggressive when a member is shutting down.
+   *
+   * Accesses to this map should be synchronized on the map to avoid concurrent
+   * modification exceptions
```

Comment:
Since we only ever `shutdownMembers.put(id, id)` this can be a `Set<ID>`. It needn't be a map.

Furthermore, if we made it a concurrent map-based set, we could eliminate all the synchronization below.

This declaration could become:

```
private final Set<ID> shutdownMembers = ConcurrentHashMap.newKeySet();
```

Then code below like this:

```
    synchronized (shutdownMembers) {
      if (!shutdownMembers.containsKey(dm)) {
        // if we've received a shutdown message then DistributionManager will already have
        // notified listeners
        listener.memberDeparted(dm, crashed, reason);
      }
    }

```

changes to this:

```
      if (!shutdownMembers.containsKey(dm)) {
        // if we've received a shutdown message then DistributionManager will already have
        // notified listeners
        listener.memberDeparted(dm, crashed, reason);
      }
```

and code below like this:

```
    synchronized (shutdownMembers) {
      return latestView.getMembers().stream().filter(id -> !shutdownMembers.containsKey(id))
          .collect(
              Collectors.toSet());
    }
```

changes to code like this:

```
    final Set<ID> membersNotShuttingDown = new HashSet<ID>(latestView.getMembers());
    final Set<ID> shutdownMembersStable = this.shutdownMembers;
    membersNotShuttingDown.removeIf(shutdownMembersStable::contains);
    return membersNotShuttingDown;
```

Reply:
@bschuchardt points out that this won't work by itself since we'd lose the bounding behavior of the `BoundedLinkedHashMap`. It's ensuring that the collection never holds more than 1000 items. And it's the 1000 most-recent ones.

Reply:
We could keep the synchronization and improve things a tiny bit by declaring the field this way:

```
  private final Set<ID> shutdownMembers = new BoundedLinkedHashMap<ID,ID>().keySet();
```

That'd let us `shutdownMembers.put(id)` and use `shutdownMembers.contains(id)` instead of `shutdownMembers.containsKey(id)`.